Source/OpenPF2Core/Private/Abilities/Tasks/PF2AbilityTask_WaitForInitiativeTurn.cpp

```diff
@@ -0,0 +1,359 @@
+			)
+		);
+	}
+}
+
+void UPF2AbilityTask_WaitForInitiativeTurn::Activate_Server(IPF2CharacterInterface* PF2Character)
+{
+	const UWorld* const World = this->GetWorld();
+
+	UE_LOG(
+		LogPf2CoreEncounters,
+		VeryVerbose,
+		TEXT("[%s] Activate_Server() called on ability task ('%s')."),
+		*(PF2LogUtilities::GetHostNetId(this->GetWorld())),
+		*(this->GetName())
+	);
+
+	if (World != nullptr)
+	{
+		IPF2GameModeInterface* PF2GameMode = Cast<IPF2GameModeInterface>(World->GetAuthGameMode());
+
+		if (PF2GameMode != nullptr)
+		{
+			TScriptInterface<IPF2CharacterInterface> CharacterScriptInterface =
+				PF2InterfaceUtilities::ToScriptInterface<IPF2CharacterInterface>(PF2Character);
+
+			TScriptInterface<IPF2QueuedActionInterface> ThisScriptInterface =
+				PF2InterfaceUtilities::ToScriptInterface<IPF2QueuedActionInterface>(this);
+
+			EPF2ActionQueueResult QueueResult;
+
+			this->GameMode = PF2GameMode;
+
+			PF2GameMode->QueueActionForInitiativeTurn(CharacterScriptInterface, ThisScriptInterface, QueueResult);
+
+			UE_LOG(
+				LogPf2CoreEncounters,
+				VeryVerbose,
+				TEXT("[%s] Result of queueing action ('%s') was: '%s'"),
+				*(PF2LogUtilities::GetHostNetId(this->GetWorld())),
+				*(this->GetIdForLogs()),
+				*(PF2EnumUtilities::ToString(QueueResult))
+			);
+
+			switch (QueueResult)
+			{
+			case EPF2ActionQueueResult::Refused:
+				// The MoPRS refused to allow this action at all (rare).
+				this->CancelAction();
+				break;
+
+			case EPF2ActionQueueResult::ExecutedImmediately:
+				// The MoPRS already executed the action without any queueing, so we have nothing left to do.
+				break;
+
+			default:
+			case EPF2ActionQueueResult::Queued:
+				IPF2GameplayAbilityInterface* Pf2Ability = this->GetOwningPf2Ability();
+
+				if (Pf2Ability != nullptr)
+				{
+					// The MoPRS queued the action for later execution. Give the ability the chance to temporarily
+					// unblock other abilities that would otherwise be incompatible with it.
+					Pf2Ability->OnQueued();
+				}
+
+				this->OnQueued.Broadcast();
+				this->SetWaitingOnRemotePlayerData();
+				break;
+			}
+		}
+	}
+}
+
+void UPF2AbilityTask_WaitForInitiativeTurn::OnPerformAction_Client()
+{
+	if (this->HasAsc())
+	{
+		this->AbilitySystemComponent->ConsumeGenericReplicatedEvent(
+			EAbilityGenericReplicatedEvent::GenericSignalFromServer,
+			this->GetAbilitySpecHandle(),
+			this->GetActivationPredictionKey()
+		);
+	}
+
+	this->PerformAction();
+}
+
+bool UPF2AbilityTask_WaitForInitiativeTurn::CanAbilityProceed() const
+{
+	bool                             bCanProceed;
+	const FGameplayAbilitySpecHandle AbilitySpecHandle = this->Ability->GetCurrentAbilitySpecHandle();
+	const FGameplayAbilityActorInfo  ActorInfo         = this->Ability->GetActorInfo();
+	IPF2GameplayAbilityInterface*    Pf2Ability        = this->GetOwningPf2Ability();
+
+	if (Pf2Ability != nullptr)
+	{
+		// Prevent blocking ourselves.
+		Pf2Ability->ForceSuspendBlocking();
```

Comment:
What about costs?

Reply:
If two queued actions both block each other with the `GameplayAbility.Type.QueueableAction` tag, I don't believe that we need to worry about suspending cost enforcement here. Actions get evaluated in the order they were queued, so the only way that action points will get spent is by them getting unblocked and the tags ensure that they fire in sequence instead of concurrently.

If I'm wrong, we can revisit this later.